### PR TITLE
Anonymize sensitive data in queries exposed to RestSqlAction logs

### PR DESCRIPTION
**Pull request by chloe-zh**
*Issue #, if available:*

* [**#97 RestSqlAction logs raw SQL queries during normal execution flow**](https://github.com/opendistro-for-elasticsearch/sql/issues/97)

*Description of changes:*

* Created a rewriting sensitive data rule
* Created a sensitive data anonymizer
* Changed the log message for the `incoming request` to a request with sensitive data gone
* Added UT

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/QueryDataMask.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import com.alibaba.druid.sql.ast.expr.SQLQueryExpr;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.identifier.RemoveSensitiveDataRule;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.amazon.opendistroforelasticsearch.sql.utils.Util.toSqlExpr;
+
+/**
+ * Utility class for mask sensitive info in incoming sql queries
```
Comment (abbashus) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/QueryDataMask.java:26]: np:
`for` -> `to`
`info` -> `information`
`sql` -> `SQL`

Same elsewhere.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java [hunk 2]
```diff
@@ -105,4 +107,15 @@ public BoolQueryBuilder checkAndAddFilter(BoolQueryBuilder boolQuery) throws Sql
...
     }
+
+    public String removeSensitiveData() {
+        String query = "";
+        try {
+            query = QueryDataMask.maskData(this.sql);
+        } catch (Exception e) {
+            LogManager.getLogger().error("Caught an exception when removing sensitive data: "
+                    + e.getClass().getSimpleName());
+        }
```
Comment (abbashus) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java:118]: Consider using 

 ```
private static final Logger LOG = LogManager.getLogger(SqlRequest.class);

...

 LOG.error("Caught an exception when masking sensitive data ",  e)

```

Let the stack trace be logged for faster debugging, instead of Exception name.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataMaskTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        String expectedQuery = "SELECT *** FROM ***";
+        Assert.assertEquals(maskedQuery, expectedQuery);
+    }
+
+    @Test
+    public void selectQueriesWithQuotedIdentifiersWithSpacesShouldMaskSensitiveData() {
+        String maskedQuery = QueryDataMask.maskData("SELECT `a   g e` FROM accounts");
+        String expectedQuery = "SELECT *** FROM ***";
+        Assert.assertEquals(maskedQuery, expectedQuery);
+    }
```
Comment (abbashus) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataMaskTest.java:50]: Does this only works for simple queries?
Can you add tests for nested, subquery, join and union/minus queries as well for better coverage.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataMaskTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        String maskedQuery = QueryDataMask.maskData("SELECT age FROM accounts");
+        String expectedQuery = "SELECT *** FROM ***";
+        Assert.assertEquals(maskedQuery, expectedQuery);
+    }
+
+    @Test
+    public void selectQueriesShouldMaskAllSensitiveData() {
+        String query = "SELECT ABS(balance), lastname FROM accounts WHERE age > 30";
+        String maskedQuery = QueryDataMask.maskData(query);
+        Assert.assertEquals(maskedQuery, "SELECT ABS(***), *** FROM *** WHERE *** > 30");
```
Comment (penghuo) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataMaskTest.java:35]: Question, what is definition sensitive data? age or 30?
IMO, the value is more sensitive than field. e.g. password = '123456', it more make sense to mask the value like password = 'xxx'

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/StringUtils.java [hunk 5]
```diff
@@ -126,6 +126,11 @@ public static boolean isNumeric(String text) {
         return Doubles.tryParse(text) != null;
     }
 
+    public static String getFirstWord(String sql) {
```
Comment (penghuo) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/StringUtils.java:129]: add UT

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java [hunk 6]
```diff
@@ -112,7 +112,8 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
             }
 
             final SqlRequest sqlRequest = SqlRequestFactory.getSqlRequest(request);
-            LOG.info("[{}] Incoming request {}: {}", LogUtils.getRequestId(), request.uri(), sqlRequest.getSql());
+            LOG.info("[{}] Incoming request {}: {}", LogUtils.getRequestId(), request.uri(),
+                    sqlRequest.removeSensitiveData());
```
Comment (penghuo) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java:116]: why adding a method on SqlRequest? could we use QueryDataMask.mask(sqlRequest.sql) instead.

**Comment by penghuo** [comment]
1. Do we need the setting to control this feature?
2. How this solution could be used in other place when logging is needed?

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataMaskTest.java (hunk 4 above)
Comment (dai-chen) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataMaskTest.java:35]: Agree. I think we can just mask all values (rewrite literals, number, string etc.)

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/QueryDataMask.java (hunk 1 above)
Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/QueryDataMask.java:26]: Will do, thanks!

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java (hunk 2 above)
Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java:118]: Sounds good, will do. Thanks!

**Comment by abbashus** [comment]
Open to discussion:

I assume more work, but instead of masking, why not anonymize the identifiers and literals. This way we will not loose the semantics.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/RemoveSensitiveDataRule.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    public boolean match(SQLQueryExpr expr) {
+        return true;
+    }
+
+    @Override
+    public void rewrite(SQLQueryExpr expr) {
+        expr.accept(this);
+    }
+
+    public Set<String> getIdentifierSet() {
```
Comment (abbashus) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/RemoveSensitiveDataRule.java:45]: np: `getIdentifierSet ` -> `getDataSet `?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java [hunk 8]
```diff
@@ -105,4 +109,14 @@ public BoolQueryBuilder checkAndAddFilter(BoolQueryBuilder boolQuery) throws Sql
         }
         return boolQuery;
     }
+
+    public String removeSensitiveData() {
+        String query = "";
+        try {
+            query = QueryDataMask.maskData(this.sql);
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java:116]: As Peng commented, I think you can either move all mask logic here or just call `maskData()` outside along with `getSql()`.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/RemoveSensitiveDataRule.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+public class RemoveSensitiveDataRule extends MySqlASTVisitorAdapter implements RewriteRule<SQLQueryExpr> {
+
+    private Set<String> dataSet = new HashSet<>();
+
+    @Override
+    public boolean visit(SQLIdentifierExpr identifierExpr) {
+        dataSet.add(identifierExpr.getName());
+        return true;
+    }
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/RemoveSensitiveDataRule.java:33]: Because values (literals) in query is more sensitive, you may need to rewrite methods like `visit(SQLNumberExpr x)`, `visit(SQLCharExpr x)` etc. We can add a few UTs to cover common literals such as string, numerical, bool, datetime etc.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/QueryDataMask.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+import com.amazon.opendistroforelasticsearch.sql.rewriter.identifier.RemoveSensitiveDataRule;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.amazon.opendistroforelasticsearch.sql.utils.Util.toSqlExpr;
+
+/**
+ * Utility class to mask sensitive information in incoming SQL queries
+ */
+public class QueryDataMask {
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/QueryDataMask.java:28]: I think you can move all logic in this class into `RemoveSensitiveDataRule` except rebuilding. (rename it to `MaskSensitiveDataRule`?) Then you just need a util method for rebuilding AST (after applied our masking rule) to a query string.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java [hunk 11]
```diff
@@ -37,7 +37,6 @@
     String sql;
     JSONObject jsonContent;
 
-
```
Comment (penghuo) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequest.java:40]: unnecessary change

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataAnonymizerTest.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    @Test
+    public void queriesShouldHaveAnonymousFieldAndIndex() {
+        String query = "SELECT ABS(balance) FROM accounts WHERE age > 30 GROUP BY ABS(balance)";
+        String expectedQuery = "(\n" +
+                "\tSELECT ABS(identifier)\n" +
+                "\tFROM table\n" +
+                "\tWHERE identifier > number\n" +
+                "\tGROUP BY ABS(identifier)\n" +
+                "\t)";
+        Assert.assertEquals(expectedQuery, QueryDataAnonymizer.anonymizeData(query));
```
Comment (penghuo) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataAnonymizerTest.java:33]: Curious what will be the output in log4j, does it output multiple line for the same query?
If so, I would suggest to remove the newline to make it in same line. Because usually, we process the log data by using line based tools and group the data with request id. e.g., grep requestId=xxx which will return all the log corresponding to specific request.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/AnonymizeSensitiveDataRule.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+import com.alibaba.druid.sql.ast.expr.SQLCharExpr;
+import com.alibaba.druid.sql.ast.expr.SQLIdentifierExpr;
+import com.alibaba.druid.sql.ast.expr.SQLIntegerExpr;
+import com.alibaba.druid.sql.ast.expr.SQLNumberExpr;
+import com.alibaba.druid.sql.ast.expr.SQLQueryExpr;
+import com.alibaba.druid.sql.ast.statement.SQLExprTableSource;
+import com.alibaba.druid.sql.dialect.mysql.visitor.MySqlASTVisitorAdapter;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.RewriteRule;
+
+public class AnonymizeSensitiveDataRule extends MySqlASTVisitorAdapter implements RewriteRule<SQLQueryExpr> {
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/AnonymizeSensitiveDataRule.java:28]: Please add some doc here.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataAnonymizerTest.java (hunk 12 above)
Comment (chloe-zh) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/utils/QueryDataAnonymizerTest.java:33]: yes, it outputs multiple lines and tabs in the front of every line as well. ok that makes sense, i will remove the redundant returns and tabs.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/AnonymizeSensitiveDataRule.java (hunk 13 above)
Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/identifier/AnonymizeSensitiveDataRule.java:28]: Will do, thanks!
